clippy_lints/src/methods/readonly_write_lock.rs

```diff
@@ -0,0 +1,51 @@
+use super::READONLY_WRITE_LOCK;
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::mir::{enclosing_mir, visit_local_usage};
+use clippy_utils::source::snippet;
+use clippy_utils::ty::is_type_diagnostic_item;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, Node};
+use rustc_lint::LateContext;
+use rustc_middle::mir::{Location, START_BLOCK};
+use rustc_span::sym;
+
+fn is_unwrap_call(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
+    if let ExprKind::MethodCall(path, receiver, ..) = expr.kind
+        && path.ident.name == sym::unwrap
+    {
+        is_type_diagnostic_item(cx, cx.typeck_results().expr_ty(receiver).peel_refs(), sym::Result)
+    } else {
+        false
+    }
+}
+
+pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>, receiver: &Expr<'_>) {
+    if is_type_diagnostic_item(cx, cx.typeck_results().expr_ty(receiver).peel_refs(), sym::RwLock)
+        && let Node::Expr(unwrap_call_expr) = cx.tcx.hir().get_parent(expr.hir_id)
+        && is_unwrap_call(cx, unwrap_call_expr)
+        && let parent = cx.tcx.hir().get_parent(unwrap_call_expr.hir_id)
+        && let Node::Local(local) = parent
+        && let Some(mir) = enclosing_mir(cx.tcx, expr.hir_id)
+        && let Some((local, _)) = mir.local_decls.iter_enumerated().find(|(_, decl)| {
+            local.span.contains(decl.source_info.span)
+        })
+        && let Some(usage) = visit_local_usage(&[local], mir, Location {
+            block: START_BLOCK,
+            statement_index: 0,
+        })
+    {
+        let writer_never_mutated = usage[0].local_consume_or_mutate_locs.is_empty();
```

Comment:
`visit_local_usage` always returns a vec with the same length as the slice of locals we give it, so it would be weird if this indexing failed (one entry for each local), and I want to say that ICEing is probably better because it would be a bug either way (and delaying it/silently not linting is probably worse because we would possibly never know?).
Bur if you disagree, I'm also fine with doing this change

Reply:
IMO I want to avoid ICE as much as possible for our users.

But it's okay, it looks good as it is. It's because this lint's category is nursery, allow-by-default. Could we leave a comment that that is the intent?

Reply:
I don't think the lint category would actually change anything or prevent an ICE here. This code runs regardless of if the lint is allowed or denied (but it wouldn't display the diagnostics if the lint is allowed).
But okay, fair, I will change it to not use indexing. It doesn't really matter I guess, since this can never happen anyway.